File: programmers/greedy/jisu/큰수만들기.py
# 효율성 실패
from collections import deque


def solution(number, k):
    l = list(number)
    cnt = 0
    while True:
        if cnt == k:
            break
        for num in l:
            if num == "9":
                continue
            l2 = l[:]
            l2.remove(num)
            if num == "1" or "".join(l) < "".join(l2)+"0":
                l = l2[:]
                cnt += 1
                break
    return "".join(l)

# deque로 수정


def solution(number, k):
    num_dq = deque(number)
    res = [num_dq.popleft()]
    while num_dq:
        v = num_dq.popleft()
        while res and res[-1] < v and k != 0:
            res.pop()
            k -= 1
        res.append(v)
        while (not num_dq) and (k != 0):
            res.pop()
            k -= 1
    return ''.join(res)

File: programmers/greedy/sangwoo/큰수만들기.js
// 그리디 알고리즘은 다시 공부를 해봐야 할 필요가 있음.
// 스택을 활용해서 앞 자리에 있는 값이 뒷 자리에 값보다 작으면 제거하는 방식으로 구현
function solution(number, k) {
    var answer = [];
   
    for(let i=0;i<number.length;i++){
        let num = number[i];
        while(k>0 && answer[answer.length-1] < num){
            answer.pop(num);
            k--;
        }
         answer.push(num);   
    }
    
    if(k){
       answer= answer.slice(0,answer.length-k);
    }
    // console.log(answer.join(''));
    return answer.join('');

}

/* 테스트 10번 시간초과 코드
function solution(number, k) {
    var answer = '';
    var arr = number.split("");
    for(let i =0;i<arr.length-1;i++){
       if(Number(arr[i])*(arr.length-i) <= Number(arr[i+1])*(arr.length-(i+1))){
           // console.log(arr[i], arr[i+1]);
        arr.splice(i,1);
        i=-1;
        k--;
        }
        if(k==0)
            break;
    }
   
    if(k){
      arr =  arr.slice(0,arr.length-k);
     
    }
    return String(arr.join(''));

}

*/